components/identity/org.wso2.carbon.identity.totp/src/main/java/org/wso2/carbon/identity/totp/TOTPTokenGenerator.java

```diff
@@ -0,0 +1,234 @@
+import org.wso2.carbon.identity.notification.mgt.NotificationSender;
+import org.wso2.carbon.identity.notification.mgt.bean.PublisherEvent;
+import org.wso2.carbon.identity.totp.exception.TOTPException;
+import org.wso2.carbon.identity.totp.internal.TOTPManagerComponent;
+import org.wso2.carbon.identity.totp.util.TOTPUtil;
+import org.wso2.carbon.user.api.UserRealm;
+import org.wso2.carbon.user.api.UserStoreException;
+import org.wso2.carbon.user.api.UserStoreManager;
+import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+import java.nio.ByteBuffer;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+
+
+/**
+ * TOTP Token generator class.
+ */
+public class TOTPTokenGenerator {
+
+	private static Log log = LogFactory.getLog(TOTPTokenGenerator.class);
+	private static volatile TOTPTokenGenerator instance;
+	private final String eventName = "userOperation";
+	private final String usernameLabel = "username";
+	private final String operationLabel = "operation";
+
+	private TOTPTokenGenerator() {
+	}
+
+	;
+
+	/**
+	 * Singleton method to get instance of TOTPTokenGenerator.
+	 *
+	 * @return instance of TOTPTokenGenerator
+	 */
+	public static TOTPTokenGenerator getInstance() {
+		if (instance == null) {
+			synchronized (TOTPTokenGenerator.class) {
+				if (instance == null) {
+					instance = new TOTPTokenGenerator();
+				}
+			}
+		}
+		return instance;
+	}
+
+	/**
+	 * Generate TOTP token for a locally stored user.
+	 *
+	 * @param username username of the user
+	 * @return TOTP token as a String
+	 * @throws TOTPException
+	 */
+	public String generateTOTPTokenLocal(String username)
+			throws TOTPException {
+
+		long token = 0;
+		try {
+			int tenantId = IdentityUtil.getTenantIdOFUser(username);
+			UserRealm userRealm = TOTPManagerComponent.getRealmService().getTenantUserRealm(tenantId);
+			if (userRealm != null) {
+				UserStoreManager userStoreManager = userRealm.getUserStoreManager();
+				String secretKey = userStoreManager.getUserClaimValue(MultitenantUtils.getTenantAwareUsername
+						(username), Constants.SECRET_KEY, null);
+
+				byte[] secretkey;
+				String encoding = "Base32";
+				try {
+					encoding = TOTPUtil.getEncodingMethod();
+				} catch (IdentityApplicationManagementException e) {
+					log.error("Error when fetching the encoding method");
+				}
+
+				if ("Base32".equals(encoding)) {
+					Base32 codec32 = new Base32();
+					secretkey = codec32.decode(secretKey);
+				} else {
+					Base64 code64 = new Base64();
+					secretkey = code64.decode(secretKey);
+				}
+				try {
+					token = getCode(secretkey, getTimeIndex());
+					//sendNotification("token : " + token, username);
+					if (log.isDebugEnabled()) {
+						log.debug("Token is sent to via email. token : " + token);
+					}
+					log.info("Generated token : " + token); //space both sides of +
```

Comment:
This token is used for login right ? Since this is confidential information avoid from logging them